### PR TITLE
[PM-2135] [BEEEP] Refactor and refresh web user verification components

### PR DESCRIPTION
**Pull request by coroiu**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

- Move to it’s own module
- Re-export from `LooseComponentsModule` to minimize impact on the rest of the codebase
- Upgrade form to use CL
- Upgrade dialog to use CL
- Keep `BaseComponent` structure

## Code changes

See objective

#### Prompt with password
image

#### Prompt with OTP
image

#### Prompt with OTP Sent
image

#### Prompt with wrong password
image

#### Prompt with password
image

#### Prompt with OTP
image

#### Prompt with OTP Sent
image

#### Prompt with field error
image

#### Prompt with incorrect password
image

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: libs/angular/src/auth/components/user-verification.component.ts [hunk 1]
```diff
@@ -18,17 +19,42 @@ import { Verification } from "@bitwarden/common/types/verification";
...
+  get invalidSecret() {
+    return this._invalidSecret;
+  }
+  set invalidSecret(value: boolean) {
+    this._invalidSecret = value;
+    this.invalidSecretChange.emit(value);
+    if (value) {
+      this.secret.markAsTouched();
+    }
+    this.secret.updateValueAndValidity({ emitEvent: false });
```
Comment (coroiu) [inline libs/angular/src/auth/components/user-verification.component.ts:33]: **point-of-interest:** this is pretty damn hacky but unfortunately there is no way of knowing if the parent control has been marked as touched, see: https://github.com/angular/angular/issues/10887

File: libs/angular/src/auth/components/user-verification-prompt.component.ts [hunk 2]
```diff
@@ -27,11 +33,23 @@ export class UserVerificationPromptComponent {
...
+      return;
+    }
+
     try {
       //Incorrect secret will throw an invalid password error.
       await this.userVerificationService.verifyUser(this.secret.value);
+      this.invalidSecret = false;
     } catch (e) {
+      this.invalidSecret = true;
       this.platformUtilsService.showToast(
```
Comment (jlf0dev) [inline libs/angular/src/auth/components/user-verification-prompt.component.ts:53]: Question: Are we showing toasts along with form errors in other places as well? I feel like the form error is probably enough

File: libs/angular/src/auth/components/user-verification.component.ts (hunk 1 above)
Comment (jlf0dev) [reply libs/angular/src/auth/components/user-verification.component.ts:34]: That's unfortunate, could we add a comment here explaining ourselves?

**Review by jlf0dev — COMMENTED** [review]
Thanks for doing this! I think it's looking good, just had a couple comments.

Could you also do me a favor while you're here? We need to move the `userVerification.service.abstraction.ts` and `userVerification-api.service.abstraction.ts` to the Auth folder. If you could rename them as well that would be great. Thanks!

File: libs/angular/src/auth/components/user-verification-prompt.component.ts (hunk 2 above)
Comment (coroiu) [reply libs/angular/src/auth/components/user-verification-prompt.component.ts:53]: You have a good point and I don;t think we are, the reason I'm still showing them is because the toast logic comes from the base component and is needed to show toasts in other clients. Also the error message needs additional code to be able to be displayed which is not yet implemented everywhere. I started migrating all other dialogs that were using user verification as well but realized how much work it would be and gave up 😅

File: libs/angular/src/auth/components/user-verification.component.ts (hunk 1 above)
Comment (coroiu) [reply libs/angular/src/auth/components/user-verification.component.ts:39]: Added a comment!

File: libs/angular/src/auth/components/user-verification-prompt.component.ts (hunk 2 above)
Comment (jlf0dev) [reply libs/angular/src/auth/components/user-verification-prompt.component.ts:53]: Ah, fair enough! We'll get there eventually 😄

File: libs/angular/src/auth/components/user-verification-prompt.component.ts [hunk 3]
```diff
@@ -27,11 +33,23 @@ export class UserVerificationPromptComponent {
...
+  submit = async () => {
+    this.formGroup.markAllAsTouched();
+
+    if (this.formGroup.invalid) {
+      return;
+    }
+
     try {
       //Incorrect secret will throw an invalid password error.
       await this.userVerificationService.verifyUser(this.secret.value);
```
Comment (eliykat) [inline libs/angular/src/auth/components/user-verification-prompt.component.ts:49]: Do you have any thoughts on how to improve this design? I never really liked how I implemented this, it's basically using exceptions for code flow which is generally considered an anti-pattern. We shouldn't have to wrap the `userVerificationService.verifyUser` call in a try/catch block as a matter of course.

Maybe it should return a `{ success: boolean, reason: string}` object instead, or maybe the interface just needs rethinking. e.g. it does seem like the "Master password/verification code required" error message should be handled by a simple form control validator.

File: libs/angular/src/auth/components/user-verification-prompt.component.ts [hunk 4]
```diff
@@ -27,11 +33,23 @@ export class UserVerificationPromptComponent {
...
+      this.invalidSecret = false;
     } catch (e) {
+      this.invalidSecret = true;
       this.platformUtilsService.showToast(
         "error",
         this.i18nService.t("error"),
```
Comment (eliykat) [inline libs/angular/src/auth/components/user-verification-prompt.component.ts:55]: The next line (which Github isn't letting me comment on) is incorrect, it should be using `e.message` which is set by `userVerificationService.verifyUser` to be an i18n'd string describing the reason for failure. Not your work I know, but while we're here...

File: libs/components/src/form-field/form-field.component.ts [hunk 5]
```diff
@@ -30,6 +30,19 @@ export class BitFormFieldComponent implements AfterContentChecked {
...
+    this._disableMargin = coerceBooleanProperty(value);
+  }
+  get disableMargin() {
+    return this._disableMargin;
+  }
+
+  @HostBinding("class")
+  get classList() {
+    return ["tw-block"].concat(this.disableMargin ? [] : ["tw-mb-6"]);
+  }
```
Comment (eliykat) [inline libs/components/src/form-field/form-field.component.ts:44]: It looks like we're disabling the bottom margin. Why is this required - what's the visual impact if we don't?

File: libs/angular/src/auth/components/user-verification.component.ts [hunk 6]
```diff
@@ -17,29 +19,61 @@ import { Verification } from "@bitwarden/common/types/verification";
...
   disableRequestOTP = false;
   sentCode = false;
 
-  secret = new FormControl("");
+  secret = new FormControl("", [
+    Validators.required,
+    () => {
+      if (this.invalidSecret) {
+        return {
+          invalidSecret: { message: this.i18nService.t("incorrectPassword") },
```
Comment (eliykat) [inline libs/angular/src/auth/components/user-verification.component.ts:52]: This won't always be a password, it could be an incorrect OTP code. `verificationService` will tell you in the error it throws (see previous comment). Again, this feels like a problem with how these components are structured, it's pretty weird that we need to tell this form control whether it's valid or not.

**Review by eliykat — COMMENTED** [review]
A nice tidy up 🧹  thank you! My more general questions below are non-blocking I'd but would like your thoughts seeing as how you've spent some time here.

**Comment by coroiu** [comment]
PR temporarily paused while we wait for https://github.com/bitwarden/clients/pull/5007 to get merged (will cause conflicts)

File: libs/angular/src/auth/components/user-verification-prompt.component.ts (hunk 4 above)
Comment (coroiu) [reply libs/angular/src/auth/components/user-verification-prompt.component.ts:55]: > Not your work I know, but while we're here...

I seem to have a talent for stumbling upon these, I'll take a look as soon as the PR gets unblocked!

File: libs/components/src/form-field/form-field.component.ts (hunk 5 above)
Comment (coroiu) [reply libs/components/src/form-field/form-field.component.ts:44]: The dialog gets a super big margin at the bottom because this margin gets added to the dialogs own inner padding.

File: libs/angular/src/auth/components/user-verification-prompt.component.ts (hunk 3 above)
Comment (coroiu) [reply libs/angular/src/auth/components/user-verification-prompt.component.ts:49]: @eliykat I like this discussion, but I'm thinking it's a slightly out-of-scope for this PR, what do you think?

File: libs/angular/src/auth/components/user-verification-prompt.component.ts (hunk 4 above)
Comment (coroiu) [reply libs/angular/src/auth/components/user-verification-prompt.component.ts:55]: Fixed!

File: libs/angular/src/auth/components/user-verification-prompt.component.ts (hunk 3 above)
Comment (eliykat) [reply libs/angular/src/auth/components/user-verification-prompt.component.ts:49]: I agree, we don't have to action it here.

File: libs/angular/src/auth/components/user-verification.component.ts (hunk 6 above)
Comment (eliykat) [reply libs/angular/src/auth/components/user-verification.component.ts]: I think it's the same problem we hit on above, i.e. these components are probably using the wrong abstraction. But this fix works for now.

**Review by eliykat — APPROVED** [review]
Fixed that whitelist issue for you ;) It's just diffing 2 lists of filenames, so it'll fail if you _don't_ have a file whitelisted that should be, but also if you _do_ have a file whitelisted that shouldn't be. You still had the `userVerification` folder in there.

**Review by eliykat — APPROVED** [review]
Sorry for missing your review request!

**Review by eliykat — DISMISSED** [review]

**Review by eliykat — APPROVED** [review]

**Review by jlf0dev — DISMISSED** [review]
This looks good to me as well 👍

**Review by djsmith85 — APPROVED** [review]
Approving for team-tools-dev and on behalf of team-leads-eng

File: apps/web/src/app/auth/settings/deauthorize-sessions.component.ts [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import { Component } from "@angular/core";
 
 import { ApiService } from "@bitwarden/common/abstractions/api.service";
-import { UserVerificationService } from "@bitwarden/common/abstractions/userVerification/userVerification.service.abstraction";
+import { UserVerificationService } from "@bitwarden/common/auth/abstractions/user-verification/user-verification.service.abstraction";
```
Comment (Hinton) [inline apps/web/src/app/auth/settings/deauthorize-sessions.component.ts:4]: Would have been nice to append abstraction to that service with this change since we had to update each file either way 😭

File: apps/web/src/app/shared/loose-components.module.ts [hunk 8]
```diff
@@ -230,6 +230,7 @@ import { SharedModule } from "./shared.module";
     LowKdfComponent,
   ],
   exports: [
+    UserVerificationModule,
```
Comment (Hinton) [inline apps/web/src/app/shared/loose-components.module.ts:233]: Seem we only export it? Could we move the imports to wherever they are being used.

File: apps/web/src/app/auth/shared/components/user-verification/user-verification.component.html [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    <bit-hint>{{ "confirmIdentity" | i18n }}</bit-hint>
+  </bit-form-field>
+</ng-container>
+<ng-container *ngIf="usesKeyConnector">
+  <div class="tw-mb-6">
+    <label class="tw-block">{{ "sendVerificationCode" | i18n }}</label>
+    <button type="button" bitButton buttonType="secondary" [bitAction]="requestOTP" appAutofocus>
+      {{ "sendCode" | i18n }}
+    </button>
+    <span class="tw-ml-2 tw-text-success" role="alert" @sent *ngIf="sentCode">
```
Comment (Hinton) [inline apps/web/src/app/auth/shared/components/user-verification/user-verification.component.html:23]: What does `@sent` do?

File: libs/angular/src/auth/components/user-verification.component.ts [hunk 10]
```diff
@@ -17,29 +19,65 @@ import { Verification } from "@bitwarden/common/types/verification";
...
+      if (this.invalidSecret) {
+        return {
+          invalidSecret: {
+            message: this.usesKeyConnector
+              ? this.i18nService.t("incorrectCode")
+              : this.i18nService.t("incorrectPassword"),
+          },
+        };
+      }
+    },
```
Comment (Hinton) [inline libs/angular/src/auth/components/user-verification.component.ts:59]: This feels a bit odd, wouldn't https://angular.io/api/forms/AbstractControl#setErrors seem more reasonable?

**Review by Hinton — APPROVED** [review]
Some optional improvements.

**Review by willmartian — APPROVED** [review]
Approved for `team-secrets-manager-dev` 👍

**Review by eliykat — APPROVED** [review]
I've already reviewed but just re-approving on behalf of Billing Team given that they're still part of AC.

File: apps/web/src/app/shared/loose-components.module.ts (hunk 8 above)
Comment (coroiu) [reply apps/web/src/app/shared/loose-components.module.ts:233]: I don't think we granular modules like that, I think I tried

File: apps/web/src/app/auth/settings/deauthorize-sessions.component.ts (hunk 7 above)
Comment (coroiu) [reply apps/web/src/app/auth/settings/deauthorize-sessions.component.ts:4]: Oh, didn't know that was a convention

File: libs/angular/src/auth/components/user-verification.component.ts (hunk 10 above)
Comment (coroiu) [reply libs/angular/src/auth/components/user-verification.component.ts:59]: Oh I tried :( `setErrors` completely overwrites all errors and can then be cleared by other validators running

File: apps/web/src/app/auth/shared/components/user-verification/user-verification-prompt.component.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+ */
+export const openUserVerificationPrompt = (
+  dialogService: DialogServiceAbstraction,
+  config: DialogConfig<UserVerificationPromptParams>
+) => {
+  return dialogService.open<boolean, UserVerificationPromptParams>(
+    UserVerificationPromptComponent,
+    config
+  );
+};
```
[the remark this answers is not in the document]
Comment (coroiu) [reply apps/web/src/app/auth/shared/components/user-verification/user-verification-prompt.component.ts:60]: Multiple conventions as always :D If I remember correctly this convention was originally created by @shane-melton as part of the OAVR work in EC-pod

**Review by jlf0dev — APPROVED** [review]
Approving for `team-auth-dev`

File: apps/web/src/app/auth/shared/components/user-verification/user-verification-prompt.component.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (shane-melton) [reply apps/web/src/app/auth/shared/components/user-verification/user-verification-prompt.component.ts:60]: Yep that was me, I went with the separate exported method originally; but, I'm not attached to that convention by any means. A static `open` method on the class itself makes sense too and am open to standardizing on that convention instead if that's preferred.

File: apps/web/src/app/auth/shared/components/user-verification/user-verification.component.html (hunk 9 above)
Comment (willmartian) [reply apps/web/src/app/auth/shared/components/user-verification/user-verification.component.html:23]: @Hinton @coroiu fyi, it is an animation trigger: https://github.com/bitwarden/clients/blob/master/apps/web/src/app/auth/shared/components/user-verification/user-verification.component.ts#L18
